cmd/krel/cmd/changelog_test.go

```diff
@@ -17,12 +17,157 @@ limitations under the License.
 package cmd
 
 import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
 	"testing"
 
 	"github.com/stretchr/testify/require"
+	"k8s.io/release/pkg/command"
+	"k8s.io/release/pkg/git"
 )
 
 func TestChangelogNoArgumentsOrFlags(t *testing.T) {
 	err := runChangelog()
 	require.NotNil(t, err)
 }
+
+type sut struct {
+	repo        *git.Repo
+	bareRepoDir string
+	tempDir     string
+}
+
+func generateSSHKeyIfNeeded(t *testing.T) {
+	sshDir := filepath.Join(os.Getenv("HOME"), ".ssh")
+	require.Nil(t, os.MkdirAll(sshDir, 0o600))
+
+	sshKeyFile := filepath.Join(sshDir, "id_rsa")
+	if _, err := os.Stat(sshKeyFile); os.IsNotExist(err) {
+		require.Nil(t, command.New(
+			"ssh-keygen", "-b", "2048", "-f", sshKeyFile, "-t", "rsa", "-q", "-N", "",
+		).RunSuccess())
+	}
+}
+
+func newSUT(t *testing.T, replayDir string) *sut {
+	// A local bare k/k repo will be our immutable test base
+	cwd, err := os.Getwd()
+	require.Nil(t, err)
+	testDataDir := filepath.Join(cwd, "test")
+	tempDir := filepath.Join(testDataDir, "tmp")
+	require.Nil(t, os.MkdirAll(tempDir, 0o755))
+
+	baseBareDir := filepath.Join(tempDir, "k8s-test-changelog-bare-testbase")
+	if _, err := os.Stat(baseBareDir); os.IsNotExist(err) {
+		require.Nil(t, command.New(
+			"git", "clone", "--bare",
+			fmt.Sprintf(
+				"https://github.com/%s/%s",
+				git.DefaultGithubOrg,
+				git.DefaultGithubRepo,
+			),
+			baseBareDir,
+		).RunSuccess())
+	}
+
+	// Now creaete the bare repo from it which can be modified by the test
+	bareRepoDir, err := ioutil.TempDir(tempDir, "k8s-test-changelog-bare-")
+	require.Nil(t, err)
+	require.Nil(t, command.New(
+		"git", "clone", "--bare", baseBareDir, bareRepoDir,
+	).RunSuccess())
+
+	// Now the system under test, which is a clone from the bare repo above
+	generateSSHKeyIfNeeded(t)
+	sutDir, err := ioutil.TempDir(tempDir, "k8s-test-changelog-sut-")
+	require.Nil(t, err)
+	require.Nil(t, os.RemoveAll(sutDir))
+	repo, err := git.CloneOrOpenRepo(sutDir, bareRepoDir, true)
+	require.Nil(t, err)
```

Comment:
Can you please add some comments or so on why we need 3 (or 2?) repos:
- a (bare) clone of k/k on 'k8s-test-changelog-bare-testbase*'
- a (bare) clone of the above in 'k8s-test-changelog-bare-*'
- a "regular" clone of the above in 'k8s-test-changelog-sut-*'

Again, this all has probably good reasons. Making that easier to discover and understand would be highly appreciated!

Reply:
I think I can simplify this since we do not push any more with `krel changelog`. :)